File: Assignment2/python_solution/utils/enumerated_object.py
""" Base class definition for objects that should have enumerated instances. """

from collections import defaultdict

from utils import type_of


class EnumeratedObject:
    """
    Base class that will enumerate all instances of a sub-class.

    Each unique sub-class gets a unique enumeration.
    """

    counts = defaultdict(int)

    def __init__(self):
        cls = type(self)
        self.id = EnumeratedObject.counts[cls] + 1
        EnumeratedObject.counts[cls] += 1

    def __str__(self):
        return f"{type_of(self)} ID: {self.id}"

    def __hash__(self):
        return self.id

    def __eq__(self, other):
        if not isinstance(other, type(self)):
            return False

        return self.id == other.id

File: Assignment2/cpp_source/utils/enumerated_object.cpp

/*
    Author: Jordan Cottle
    Created: 10/02/2020

    Base struct definition for objects that should have enumerated instances.
*/

#include <unordered_map>
#include <string>
#include <iostream>

#include "misc.h"
#include "enumerated_object.h"

EnumeratedObject::EnumeratedObject(std::string type)
{
    this->type = type;

    this->id = EnumeratedObject::counts[type] + 1;

    EnumeratedObject::counts[type] = this->id;
}

bool EnumeratedObject::operator==(const EnumeratedObject &other)
{
    return this->id == other.id;
}

std::string EnumeratedObject::__str__()
{
    return this->type + " object " + str(this->id);
}

std::unordered_map<std::string, int> EnumeratedObject::counts = std::unordered_map<std::string, int>();

namespace std
{
    template <>
    struct hash<EnumeratedObject>
    {
        std::size_t operator()(const EnumeratedObject &object) const
        {
            return (std::size_t)object.id;
        }
    };

} // namespace std